apps/examples/wifi_manager_sample/wm_test_interop.c

```diff
@@ -0,0 +1,662 @@
+
+	if (ap_info == NULL) {
+		WT_LOGE(TAG, "AP Info is NULL");
+		return -1;
+	}
+
+	snprintf(buf, WIFI_INTEROP_BUFSIZE, "%s" DELIMITER "%s" DELIMITER "%s" DELIMITER "\n",
+			ap_info->ssid, ap_info->security, ap_info->password);
+
+
+	file_name = WIFI_INTEROP_PATH WIFI_INTEROP_FILENAME;
+
+	ret = wm_interop_store_file(buf, WIFI_INTEROP_BUFSIZE, file_name);
+	if (ret < 0) {
+		WT_LOGE(TAG, "store file failed");
+		return -2;
+	}
+
+	WT_LEAVE;
+	return 0;
+}
+
+void  wm_interop_add_ap_config(void *arg)
+{
+	WT_ENTER;
+
+	struct wt_options *ap_info = (struct wt_options *)arg;
+	int ret = -1;
+
+	WT_LOG(TAG, "ssid[%s] security[%s] password[%s]",
+			ap_info->ssid, ap_info->security, ap_info->password);
+
+	ret = wm_interop_save_ap_config(ap_info);
+	if (ret < 0) {
+		WT_LOGE(TAG, "Save AP configuration failed");
+		return;
+	}
+
+	WT_LEAVE;
+}
+
+void build_ap_config(int idx, char *token, interop_ap_info_s *ap_info)
+{
+	switch (idx) {
+	case SSID:
+		memset(ap_info->ap_config.ssid, 0, WIFIMGR_SSID_LEN + 1);
+		ap_info->ap_config.ssid_length = strlen(token);
+		memcpy(ap_info->ap_config.ssid, token, ap_info->ap_config.ssid_length);
+
+		break;
+	case SECURITY:
+		memcpy(ap_info->security, token, MAX_INTEROP_SECURITY_LEN);
+		ap_info->ap_config.ap_auth_type = wm_get_auth_type(token);
+		if (ap_info->ap_config.ap_auth_type == WIFI_MANAGER_AUTH_UNKNOWN) {
+			WT_LOGE(TAG, "Unknown Auth type");
+			break;
+		}
+
+		if (ap_info->ap_config.ap_auth_type == WIFI_MANAGER_AUTH_OPEN
+			 || ap_info->ap_config.ap_auth_type == WIFI_MANAGER_AUTH_IBSS_OPEN) {
+
+			ap_info->ap_config.ap_crypto_type = WIFI_MANAGER_CRYPTO_NONE;
+			memset(ap_info->ap_config.passphrase, 0, WIFIMGR_PASSPHRASE_LEN + 1);
+			strcpy(ap_info->ap_config.passphrase, "");
+			ap_info->ap_config.passphrase_length = 0;
+			break;
+		}
+
+		ap_info->ap_config.ap_crypto_type = wm_get_crypto_type(token);
+		if (ap_info->ap_config.ap_crypto_type == WIFI_MANAGER_CRYPTO_UNKNOWN) {
+			WT_LOGE(TAG, "Unknown Crypto Type");
+			break;
+		}
+
+		break;
+
+	case PASSPHRASE:
+		memset(ap_info->ap_config.passphrase, 0, WIFIMGR_PASSPHRASE_LEN + 1);
+		ap_info->ap_config.passphrase_length = strlen(token);
+		memcpy(ap_info->ap_config.passphrase, token, ap_info->ap_config.passphrase_length);
+
+		if (ap_info->ap_config.ap_auth_type == WIFI_MANAGER_AUTH_WEP_SHARED) {
+			if ((strlen(token) == 13) || (strlen(token) == 26))
+				ap_info->ap_config.ap_crypto_type = WIFI_MANAGER_CRYPTO_WEP_128;
+			else if ((strlen(token) == 5) || (strlen(token) == 10))
+				ap_info->ap_config.ap_crypto_type = WIFI_MANAGER_CRYPTO_WEP_64;
+			else
+				WT_LOGE(TAG, "Unknown Crypto Type");
+		}
+
+		break;
+	default:
+		WT_LOGE(TAG, "unknown");
+		break;
+	}
+}
+
+void wifi_interop_parse_ap_config(char * buf, interop_ap_info_s *ap_info)
+{
+	char *token = NULL;
+	char *ap_buf = NULL;
+	int idx = 0;
+
+	int len = strlen(buf);
+
+	//remove the new line character '\n' from buf a,b,c,\n
+	ap_buf = strndup(buf, len-1);
```

Comment:
For `strndup`, it allocates the dynamic memory internally.
So it should be freed when it is not used anymore.
Could you check who will free this memory?
If it is used forever, it is okay.

Reply:
memory has been freed after use